main/background.js

```diff
@@ -807,381 +807,150 @@ function getInnerNfp(A, B, config){
+
+        // Update metrics
+        totalMergedLength += mergedLength;
+        fitness += calculateFitness(sheetArea, placements);
+
+        // Remove placed parts from the pool
+        rotatedParts = rotatedParts.filter(part => !placedParts.includes(part));
+
+        if (placements.length > 0) {
+            allPlacements.push({
+                sheet: sheet.source,
+                sheetId: sheet.id,
+                sheetPlacements: placements,
+            });
+        } else {
+            break; // No parts could be placed; exit loop
+        }
+    }
+
+    // Add penalty for unplaced parts
+    fitness += calculateUnplacedPenalty(rotatedParts, totalSheetArea);
+
+    // Notify completion
+    notifyProgress(nestIndex, -1);
+
+    return {
+        placements: allPlacements,
+        fitness,
+        totalMergedLength,
+    };
+}
+
+// Helper to prepare parts by rotating them
+function prepareParts(parts, rotations) {
+    return parts.map(part => {
+        let rotated = rotatePolygon(part, part.rotation);
+        rotated.rotation = part.rotation;
+        rotated.source = part.source;
+        rotated.id = part.id;
+        rotated.filename = part.filename;
+        return rotated;
+    });
+}
+
+// Place parts on a single sheet
+function placePartsOnSheet(sheet, parts, config) {
+    const placedParts = [];
+    const placements = [];
+    let mergedLength = 0;
+    let clipCache = {};
+
+    for (let i = 0; i < parts.length; i++) {
+        const part = parts[i];
+        const position = findBestPlacement(sheet, part, placements, placedParts, clipCache, config);
+
+        if (position) {
+            placedParts.push(part);
+            placements.push(position);
+
+            if (position.mergedLength) {
+                mergedLength += position.mergedLength;
+            }
+        }
+    }
+
+    return { placements, placedParts, mergedLength };
+}
+
+// Find the best placement for a single part
+function findBestPlacement(sheet, part, placements, placedParts, clipCache, config) {
+    let bestPlacement = null;
+    let minArea = Infinity;
+
+    const sheetNfp = getInnerNfp(sheet, part, config);
+    if (!sheetNfp) return null;
+
+    for (const nf of sheetNfp) {
+        for (const point of nf) {
+            const shiftVector = calculateShiftVector(point, part[0]);
+
+            const area = evaluatePlacement(sheet, part, shiftVector, placedParts, placements, config);
+            if (area < minArea) {
+                minArea = area;
+                bestPlacement = shiftVector;
+            }
+        }
+    }
+
+    return bestPlacement;
+}
+
+// Evaluate the placement quality based on bounding box or convex hull
+function evaluatePlacement(sheet, part, shiftVector, placedParts, placements, config) {
+    const shiftedPart = shiftPolygon(part, shiftVector);
+    const combinedParts = [...placedParts.map((p, i) => shiftPolygon(p, placements[i])), shiftedPart];
+
+    if (config.placementType === 'gravity') {
```

Comment:
missing box, while box has default bounds added without multiply the width with 2